### PR TITLE
🤖 auto-generate resources for Mondoo integrations

### PR DESCRIPTION
**Pull request by afiune**
~First iteration of~ a generator that create Terraform resources for Mondoo Integrations.

Most of our integrations are very similar and the only thing that changes are the `ConfigurationOptions` per integration type. This generator iterates over all possible config options and generates the following files:

1. The terraform resource (Go)
2. Tests for the new resource (Go)
3. Basic example of the new resource (HCL)
4. Importer example (Bash)
5. Documentation of the new resource (Markdown)

Note that, once the resource is generated, we run our auto-generation of documentation, which created the docs
for new resources. Additionally we generate two hybrid generated files:

1. Providers extension (Go)
2. GQL extension (Go)

These files are generated only once, they register new integrations and declare GraphQL structs needed for the new resources to work.

Here is an example of the AzureDevops integration being generated with this generator. https://github.com/mondoohq/terraform-provider-mondoo/pull/246

~Next up, we need to code the full generation so that we write directly to `internal/` and `examples/` only.~

### CODE REVIEW

**Comment by chris-rock** [comment]
Nice step to auto-generate the integrations! If I understand the code right now, we would generate it only once. Do you think we can run this in the next step as a go generate task and replace all existing integrations with that approach as well? I would like to ensure integrations do not drift away over time.

File: gen/templates/integration_resource.go.tmpl [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+	SpaceID types.String `tfsdk:"space_id"`
+
+	// integration details
+	Mrn  types.String `tfsdk:"mrn"`
+	Name types.String `tfsdk:"name"`
+
+	// {{.ResourceClassName}} options
+}
+
+func (m integration{{.ResourceClassName}}ResourceModel) GetConfigurationOptions() *mondoov1.{{.ResourceClassName}}ConfigurationOptionsInput {
```
Comment (chris-rock) [inline gen/templates/integration_resource.go.tmpl:42]: For best practices, we should also add generated comments here so that its easier to understand what the methods are doing

File: gen/gen.go [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+import (
+	"fmt"
+	"log"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+	"text/template"
+
+	"github.com/fatih/structs"
```
Comment (chris-rock) [inline gen/gen.go:15]: We should not include unmaintained go modules.

Comment (afiune) [reply gen/gen.go:15]: I couldn't find any other go module that does this. I we ok copying the code and NOT use the import?

File: gen/gen.go [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+	}
+
+	resourceDotTFTemplateFile := filepath.Join("gen", "templates", "resource.tf.tmpl")
+	resourceTFTmpl, err := template.ParseFiles(resourceDotTFTemplateFile)
+	if err != nil {
+		return err
+	}
+
+	// Ensure the output directory exists
+	outputDirPath := filepath.Join("gen", "generated")
```
Comment (czunker) [inline gen/gen.go:54]: Couldn't we create this directly in the directory the other integrations are in?

File: gen/gen.go [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+	if err := os.MkdirAll(outputDirPath, 0755); err != nil {
+		return err
+	}
+
+	i := mondoov1.ClientIntegrationConfigurationInput{}
+	m := structs.Map(i)
+
+	for k := range m {
+		// TODO we know the type and the struct associated to the type, we need
+		// to look it (the struct) and use the same `structs.Map(v)` to list all
```
Comment (czunker) [inline gen/gen.go:64]: Something reads strange to me.
Perhaps:
```suggestion
		// to look at it (the struct) and use the same `structs.Map(v)` to list all
```

File: gen/gen.go [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+		}
+
+		// Create examples/ files
+		examplesDirPath := filepath.Join(outputDirPath, "examples", fmt.Sprintf("mondoo_integration_%s", terraformResourceName))
+		// Ensure the output directory exists
+		if err := os.MkdirAll(examplesDirPath, 0755); err != nil {
+			return err
+		}
+		// Create example main.tf
+		err = os.WriteFile(filepath.Join(examplesDirPath, "main.tf"), mainDotTFTestFile(), 0644)
```
Comment (czunker) [inline gen/gen.go:115]: Same here: Could we generate this directly in the final directory?

File: gen/gen.go [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+
+		if err := resourceTFTmpl.Execute(resourceTFFile, resource); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+var matchFirstCap = regexp.MustCompile("(.)([A-Z][a-z]+)")
```
Comment (czunker) [inline gen/gen.go:135]: I'm not sure this works as expected. Wouldn't the `.` also match an upper case letter?

Comment (czunker) [reply gen/gen.go:135]: I think, I'm missing something here. 
Could you please add a comment or a test case to show how this works?

File: gen/templates/integration_resource.go.tmpl
Comment (czunker) [inline gen/templates/integration_resource.go.tmpl:1]: What would we need in this auto generate PR to also be able to generate the docs, like in https://github.com/mondoohq/terraform-provider-mondoo/pull/194/files#diff-f9f43e36f94b402f6ad590eef25f3da2adbf02034de08d64aeb834510e3159e9

Comment (afiune) [reply gen/templates/integration_resource.go.tmpl:1]: docs will be auto generated too

File: gen/gen.go (hunk 3 above)
Comment (afiune) [reply gen/gen.go:54]: not yet. still need some more work but we will ultimately write it in the right directory.

File: gen/gen.go (hunk 5 above)
Comment (afiune) [reply gen/gen.go:115]: not yet. still need some more work but we will ultimately write it in the right directory.

File: gen/gen.go (hunk 2 above)
Comment (afiune) [reply gen/gen.go:15]: Good call! 🌟

File: gen/gen.go (hunk 5 above)
Comment (czunker) [reply gen/gen.go:115]: Is this something for a follow-up or do you want to include it in this PR?

File: gen/templates/integration_resource.go.tmpl
Comment (afiune) [reply gen/templates/integration_resource.go.tmpl:1]: Done! It does it now 💯

File: gen/gen.go (hunk 2 above)
Comment (afiune) [reply gen/gen.go]: Updated to use https://github.com/go-viper/mapstructure

File: gen/gen.go (hunk 3 above)
Comment (afiune) [reply gen/gen.go]: We are creating it directly now!! 💯

File: gen/gen.go (hunk 5 above)
Comment (afiune) [reply gen/gen.go]: No longer a follow-up. This PR now creates two auto-generated integrations.

**Review by scottford-io — COMMENTED** [review]
@afiune thanks for looping me in. I like the idea of more automation here, but a few suggestions...

- Introduce a manual review gate for generated code to validate semantic correctness and maintain Terraform UX.
- Consider a “stabilization” phase where newly generated resources are marked as experimental until vetted.
- Maintain a regression test suite that exercises real integrations to catch edge cases beyond the generated tests.
- Document generator behavior and customization hooks, so future contributors know how to extend or override safely.

File: gen/gen.go [hunk 7]
```diff
@@ -0,0 +1,475 @@
...
+		return fmt.Sprintf("mondoov1.String(m.%s.ValueString())", name)
+	case StringPtrField.MondooType:
+		return fmt.Sprintf("mondoov1.NewStringPtr(mondoov1.String(m.%s.ValueString()))", name)
+	case ArrayStringPtrField.MondooType:
+		return fmt.Sprintf("ToPtr(ConvertSliceStrings(m.%s))", name)
+	}
+	return "\"unimplemented: check gen/gen.go\""
+}
+
+func (f Field) ImportConvertion(resourceClassName, fieldName string) string {
```
Comment (jaym) [inline gen/gen.go:84]: Conversion?

File: gen/gen.go [hunk 8]
```diff
@@ -0,0 +1,475 @@
...
+	return "Optional: true"
+}
+
+type tfgenTag struct {
+	Required    bool
+	Description string
+}
+
+func parseTFGenTag(tag string) tfgenTag {
+	t := tfgenTag{}
```
Comment (jaym) [inline gen/gen.go:118]: it might be better to use the functions provided in the `reflect` package
https://pkg.go.dev/reflect#StructTag.Get

File: gen/gen.go [hunk 9]
```diff
@@ -0,0 +1,475 @@
...
+		return err
+	}
+
+	var out []byte
+	if strings.HasSuffix(filePath, ".go") {
+		// format go file with gofmt
+		out, err = format.Source(buf.Bytes())
+		if err != nil {
+			return err
+			// log.Println(err)
```
Comment (jaym) [inline gen/gen.go:270]: can probably delete these lines now?

File: gen/templates/integration_resource.go.tmpl (hunk 1 above)
Comment (jaym) [reply gen/templates/integration_resource.go.tmpl:49]: I think this comment is still valid

**Review by jaym — APPROVED** [review]

File: internal/provider/provider_generated.go [hunk 10]
```diff
@@ -0,0 +1,29 @@
+// Copyright (c) Mondoo, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package provider
+
+import "github.com/hashicorp/terraform-plugin-framework/resource"
+
+var autoGeneratedResources = []func() resource.Resource{
```
Comment (jaym) [inline internal/provider/provider_generated.go:8]: just checking, this file is not autogenerated?

File: gen/gen.go (hunk 8 above)
Comment (afiune) [reply gen/gen.go:118]: I am doing that already, this function is not getting the tag from the field, this function is parsing the tag that could contain multiple values.
